Use ProcessPoolExecutor in the framework DAPI module

As proposed in the investigation of #10789 , this issue is the short term version of #10790 .

This implementation requires a change in the `dapi.py` module to use `ProcessPoolExecutor` limited to 1 worker. This way, everything should work without further changes but performance will increase as the processing will be done by another process instead of a thread (Python GUI limitations with high CPU usage).

For more information about the tasks, check the issues above.

## Checks

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)
<!-- If changes are made to any of the following components, uncomment the corresponding line 
- [x] **Integration tests** without failures for API configuration (`/wazuh-qa/tests/integration/test_api/test_config/`)
- [x] **System tests** for agent enrollment process (`/wazuh-qa/tests/system/test_cluster/test_agent_enrollment`)
- [x] **System tests** for agent info sync process in cluster (`/wazuh-qa/tests/system/test_cluster/test_agent_info_sync`)
- [x] **System tests** for agent key polling (`/wazuh-qa/tests/system/test_cluster/test_agent_key_polling`)
- [x] **System tests** for JWT invalidation (`/wazuh-qa/tests/system/test_jwt_invalidation`)
-->

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

# Issue update

After analyzing the API integration tests and doing some manual testing, we have detected some issues with this implementation:

## BrokenProcessPool error on certain requests

With the API integration tests we noticed there are a few requests that raise a `BrokenProcessPool` error consistently:

```cmd
2021/11/12 15:40:29 ERROR: Error executing API request locally: A process in the process pool was terminated abruptly while the future was running or pending.
concurrent.futures.process._RemoteTraceback: 
'''
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 384, in wait_result_broken_or_wakeup
    result_item = result_reader.recv()
  File "/var/ossec/framework/python/lib/python3.9/multiprocessing/connection.py", line 256, in recv
    return _ForkingPickler.loads(buf.getbuffer())
TypeError: __init__() missing 1 required positional argument: 'code'
'''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 285, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
2021/11/12 15:40:29 INFO: testing 172.20.0.1 "GET /agents/upgrade_result" with parameters {"agents_list": "008"} and body {} done in 0.077s: 500
```

This error immediately kills the child process, making the process pool not usable anymore. An API restart is needed to to work again. 

Further investigating this issue, it has been proven that this exception comes from the task which the child process is running. When **this task raises an exception**, it is considered an unexpected error and the child process terminates abruptly, breaking the pool.


## Shared memory

There are some variables that are loaded by the parent process and then changed by the child process. In examples like the `GET /security/config` endpoint, the API security configuration can be modified without restarting the API. This situation is problematic as that variable is not being shared between processes and if we update it and then try to get the new value, the old one will be loaded instead. This case could be seen in the API integration tests:

```cmd
Source test stage (line 683):
  - name: Get security configuration to check if PUT method worked correctly
    request:
      verify: False
      url: "{protocol:s}://{host:s}:{port:d}/security/config"
      headers:
        Authorization: "Bearer {test_login_token}"
      method: GET
    response:
      status_code: 200
      json:
        error: !anyint
        data:
          auth_token_exp_timeout: 3000
          rbac_mode: !anystr

Formatted stage:
  name: Get security configuration to check if PUT method worked correctly
  request:
    headers:
      Authorization: Bearer eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjM2NzMxNTU2LCJleHAiOjE2MzY3MzQ1NTYsInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbMV0sInJiYWNfbW9kZSI6IndoaXRlIn0.ANSj4rc_8b049IJrJwpFVh8Xoy0JDs7KVWK8tj23rxZ42KbOmN6VPFZVaVelu0DrWF-GQBmkOGzNNE29QFl7FAWhAaiAkOjZbBdx4vJmb8UnVIKRlNkLbAJ9G0bPTzJBD62iWpSu13qR_jK1Pj151UGprxsudigI8NUlgkwZbzwqQAXx
    method: GET
    url: 'https://localhost:55000/security/config'
    verify: false
  response:
    json:
      data:
        auth_token_exp_timeout: 3000
        rbac_mode: !anystr ''
      error: !anyint ''
    status_code: 200

Errors:
E   tavern.util.exceptions.TestFailError: Test 'Get security configuration to check if PUT method worked correctly' failed:
    - Key mismatch: (expected["data"]["auth_token_exp_timeout"] = '3000' (type = <class 'int'>), actual["data"]["auth_token_exp_timeout"] = '900' (type = <class 'int'>))
```

We need to decide whether we cover these errors for the short term solution or whether we should begin to work in the long term one as this implementation will require some more time and probably will not be ready for the next release.

--- a/api/api/authentication.py
+++ b/api/api/authentication.py
@@ -3,13 +3,11 @@
 # This program is a free software; you can redistribute it and/or modify it under the terms of GPLv2
 
 import asyncio
-import copy
 import logging
 import os
-
 from concurrent.futures import ThreadPoolExecutor
 from time import time
-                               
+
 from cryptography.hazmat.primitives import serialization
 from cryptography.hazmat.primitives.asymmetric import ec
 from jose import JWTError, jwt
@@ -134,7 +132,7 @@ def change_keypair():
 def get_security_conf():
     conf.security_conf.update(conf.read_yaml_config(config_file=SECURITY_CONFIG_PATH,
                                                     default_conf=conf.default_security_configuration))
-    return copy.deepcopy(conf.security_conf)
+    return conf.security_conf
 
 
 def generate_token(user_id=None, data=None, run_as=False):

--- a/api/api/controllers/experimental_controller.py
+++ b/api/api/controllers/experimental_controller.py
@@ -24,7 +24,7 @@ def check_experimental_feature_value(func):
     @wraps(func)
     def wrapper(*args, **kwargs):
         if not configuration.api_conf['experimental_features']:
-            raise_if_exc(WazuhResourceNotFound(code=1122))
+            raise_if_exc(WazuhResourceNotFound(1122))
         else:
             return func(*args, **kwargs)
     return wrapper

--- a/api/api/controllers/test/test_experimental_controller.py
+++ b/api/api/controllers/test/test_experimental_controller.py
@@ -485,6 +485,6 @@ def test_check_experimental_feature_value(mock_exc):
         pass
     with patch('api.configuration.api_conf', new={'experimental_features': False}):
         func_()
-        mock_exc.assert_called_once_with(WazuhResourceNotFound(code=1122))
+        mock_exc.assert_called_once_with(WazuhResourceNotFound(1122))
     with patch('api.configuration.api_conf', new={'experimental_features': True}):
         func_()

--- a/api/api/middlewares.py
+++ b/api/api/middlewares.py
@@ -2,7 +2,6 @@
 # Created by Wazuh, Inc. <info@wazuh.com>.
 # This program is a free software; you can redistribute it and/or modify it under the terms of GPLv2
 
-import concurrent.futures
 from json import JSONDecodeError
 from logging import getLogger
 from time import time
@@ -21,7 +20,6 @@ from wazuh.core.exception import WazuhTooManyRequests, WazuhPermissionError
 secure_headers = SecureHeaders(server="Wazuh", csp="none", xfo="DENY")
 
 logger = getLogger('wazuh-api')
-pool = concurrent.futures.ThreadPoolExecutor()
 
 
 def _cleanup_detail_field(detail):

--- a/api/api/uri_parser.py
+++ b/api/api/uri_parser.py
@@ -28,7 +28,7 @@ class APIUriParser(connexion.decorators.uri_parsing.OpenAPIURIParser):
                 q = parse_api_param(request.url, 'q')
                 if q:
                     if ';' in q:
-                        raise_if_exc(APIError(code=2009))
+                        raise_if_exc(APIError(2009))
 
             # Transform to lowercase the values for query parameter's spec.yaml enums
             lower_fields = ['component', 'configuration', 'hash', 'requirement', 'status', 'type', 'section', 'tag',

--- a/framework/wazuh/agent.py
+++ b/framework/wazuh/agent.py
@@ -970,7 +970,7 @@ def upgrade_agents(agent_list: list = None, wpk_repo: str = None, version: str =
 
                 # Upgrade error for specific agents
                 elif (error_code := 1810 + socket_error) in ERROR_CODES_UPGRADE_SOCKET:
-                    error = WazuhError(code=error_code, cmd_error=True, extra_message=agent_result['message'])
+                    error = WazuhError(error_code, cmd_error=True, extra_message=agent_result['message'])
                     result.add_failed_item(id_=str(agent_result['agent']).zfill(3), error=error)
 
                 # Upgrade error for all agents, bad request
@@ -1067,7 +1067,7 @@ def get_upgrade_result(agent_list: list = None, filters: dict = None, q: str = N
 
                 # Upgrade error for specific agents (no task in DB)
                 elif (error_code := 1810 + task_error) in ERROR_CODES_UPGRADE_SOCKET_GET_UPGRADE_RESULT:
-                    error = WazuhError(code=error_code, cmd_error=True, extra_message=task_result['message'])
+                    error = WazuhError(error_code, cmd_error=True, extra_message=task_result['message'])
                     result.add_failed_item(id_=str(task_result['agent']).zfill(3), error=error)
 
                 # Upgrade error for all agents, internal server error

--- a/framework/wazuh/core/cluster/common.py
+++ b/framework/wazuh/core/cluster/common.py
@@ -561,7 +561,7 @@ class Handler(asyncio.Protocol):
             res = await self.send_request(b'dapi_err', json.dumps(e, cls=WazuhJSONEncoder).encode())
         except Exception as e:
             self.logger.error(f"Error sending API response to local client: {e}")
-            exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+            exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                   cls=WazuhJSONEncoder).encode()
             res = await self.send_request(b'dapi_err', exc_info)
         finally:
@@ -585,7 +585,7 @@ class Handler(asyncio.Protocol):
             await self.send_request(b'sendsyn_err', json.dumps(e, cls=WazuhJSONEncoder).encode())
         except Exception as e:
             self.logger.error(f"Error sending send sync response to local client: {e}")
-            exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+            exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                   cls=WazuhJSONEncoder).encode()
             await self.send_request(b'sendsync_err', exc_info)
         finally:

--- a/framework/wazuh/core/cluster/dapi/dapi.py
+++ b/framework/wazuh/core/cluster/dapi/dapi.py
@@ -225,7 +225,7 @@ class DistributedAPI:
             raise exception.WazuhError(1017, extra_message=extra_info)
 
     @staticmethod
-    def run_local(f, f_kwargs, logger, rbac_permissions, broadcasting, nodes, current_user, is_async=False):
+    def run_local(f, f_kwargs, logger, rbac_permissions, broadcasting, nodes, current_user):
         """Run framework SDK function locally in another process."""
 
         def debug_log(logger, message):
@@ -239,17 +239,10 @@ class DistributedAPI:
         common.broadcast.set(broadcasting)
         common.cluster_nodes.set(nodes)
         common.current_user.set(current_user)
-        try:
-            data = f(**f_kwargs)
-        except Exception as e:
-            if is_async:
-                raise e
-
-            data = e
-
+        data = f(**f_kwargs)
         common.reset_context_cache()
         debug_log(logger, "Finished executing request locally")
-        return json.dumps(data, cls=c_common.WazuhJSONEncoder) if not is_async else data
+        return data
 
     async def execute_local_request(self) -> str:
         """Execute an API request locally.
@@ -275,7 +268,7 @@ class DistributedAPI:
             try:
                 if self.is_async:
                     task = self.run_local(self.f, self.f_kwargs, self.logger, self.rbac_permissions, self.broadcasting,
-                                          self.nodes, self.current_user, is_async=True)
+                                          self.nodes, self.current_user)
 
                 else:
                     loop = asyncio.get_event_loop()
@@ -289,14 +282,7 @@ class DistributedAPI:
                                                               self.broadcasting, self.nodes,
                                                               self.current_user))
                 try:
-                    if not self.is_async:
-                        data = json.loads(await asyncio.wait_for(task, timeout=timeout),
-                                          object_hook=c_common.as_wazuh_object)
-                        if isinstance(data, Exception):
-                            raise data
-                    else:
-                        data = await asyncio.wait_for(task, timeout=timeout)
-
+                    data = await asyncio.wait_for(task, timeout=timeout)
                 except asyncio.TimeoutError:
                     raise exception.WazuhInternalError(3021)
                 except OperationalError:

--- a/framework/wazuh/core/cluster/dapi/dapi.py
+++ b/framework/wazuh/core/cluster/dapi/dapi.py
@@ -28,7 +28,6 @@ from wazuh.cluster import get_node_wrapper, get_nodes_info
 from wazuh.core import common, exception
 from wazuh.core.cluster import local_client, common as c_common
 from wazuh.core.cluster.cluster import check_cluster_status
-from wazuh.core.cluster.common import WazuhJSONEncoder, as_wazuh_object
 from wazuh.core.exception import WazuhException, WazuhClusterError, WazuhError
 from wazuh.core.wazuh_socket import wazuh_sendsync
 
@@ -226,7 +225,7 @@ class DistributedAPI:
             raise exception.WazuhError(1017, extra_message=extra_info)
 
     @staticmethod
-    def run_local(f, f_kwargs, logger, rbac_permissions, broadcasting, nodes, current_user):
+    def run_local(f, f_kwargs, logger, rbac_permissions, broadcasting, nodes, current_user, is_async=False):
         """Run framework SDK function locally in another process."""
 
         def debug_log(logger, message):
@@ -243,11 +242,14 @@ class DistributedAPI:
         try:
             data = f(**f_kwargs)
         except Exception as e:
+            if is_async:
+                raise e
+
             data = e
 
         common.reset_context_cache()
         debug_log(logger, "Finished executing request locally")
-        return json.dumps(data, cls=WazuhJSONEncoder)
+        return json.dumps(data, cls=c_common.WazuhJSONEncoder) if not is_async else data
 
     async def execute_local_request(self) -> str:
         """Execute an API request locally.
@@ -273,7 +275,7 @@ class DistributedAPI:
             try:
                 if self.is_async:
                     task = self.run_local(self.f, self.f_kwargs, self.logger, self.rbac_permissions, self.broadcasting,
-                                          self.nodes, self.current_user)
+                                          self.nodes, self.current_user, is_async=True)
 
                 else:
                     loop = asyncio.get_event_loop()
@@ -287,9 +289,13 @@ class DistributedAPI:
                                                               self.broadcasting, self.nodes,
                                                               self.current_user))
                 try:
-                    data = json.loads(await asyncio.wait_for(task, timeout=timeout), object_hook=as_wazuh_object)
-                    if isinstance(data, Exception):
-                        raise data
+                    if not self.is_async:
+                        data = json.loads(await asyncio.wait_for(task, timeout=timeout),
+                                          object_hook=c_common.as_wazuh_object)
+                        if isinstance(data, Exception):
+                            raise data
+                    else:
+                        data = await asyncio.wait_for(task, timeout=timeout)
 
                 except asyncio.TimeoutError:
                     raise exception.WazuhInternalError(3021)

--- a/framework/wazuh/core/cluster/dapi/dapi.py
+++ b/framework/wazuh/core/cluster/dapi/dapi.py
@@ -10,9 +10,9 @@ import operator
 import os
 import time
 from collections import defaultdict
-from concurrent.futures import ThreadPoolExecutor
+from concurrent.futures import ProcessPoolExecutor
 from copy import copy, deepcopy
-from functools import reduce
+from functools import reduce, partial
 from operator import or_
 from typing import Callable, Dict, Tuple, List
 
@@ -31,7 +31,7 @@ from wazuh.core.cluster.cluster import check_cluster_status
 from wazuh.core.exception import WazuhException, WazuhClusterError, WazuhError
 from wazuh.core.wazuh_socket import wazuh_sendsync
 
-threadpool = ThreadPoolExecutor(max_workers=1)
+process_pool = ProcessPoolExecutor(max_workers=1)
 
 
 class DistributedAPI:
@@ -222,6 +222,25 @@ class DistributedAPI:
             }
             raise exception.WazuhError(1017, extra_message=extra_info)
 
+    @staticmethod
+    def run_local(f, f_kwargs, logger, rbac_permissions, broadcasting, nodes, current_user):
+        """Run framework SDK function locally in another process."""
+        def debug_log(logger, message):
+            if logger.name == 'wazuh-api':
+                logger.debug2(message)
+            else:
+                logger.debug(message)
+
+        debug_log(logger, "Starting to execute request locally")
+        common.rbac.set(rbac_permissions)
+        common.broadcast.set(broadcasting)
+        common.cluster_nodes.set(nodes)
+        common.current_user.set(current_user)
+        data = f(**f_kwargs)
+        common.reset_context_cache()
+        debug_log(logger, "Finished executing request locally")
+        return data
+
     async def execute_local_request(self) -> str:
         """Execute an API request locally.
 
@@ -230,19 +249,6 @@ class DistributedAPI:
         str
             JSON response.
         """
-
-        def run_local():
-            self.debug_log("Starting to execute request locally")
-            common.rbac.set(self.rbac_permissions)
-            common.broadcast.set(self.broadcasting)
-            common.cluster_nodes.set(self.nodes)
-            common.current_user.set(self.current_user)
-            common.origin_module.set(self.origin_module)
-            data = self.f(**self.f_kwargs)
-            common.reset_context_cache()
-            self.debug_log("Finished executing request locally")
-            return data
-
         try:
             if self.f_kwargs.get('agent_list') == '*':
                 del self.f_kwargs['agent_list']
@@ -256,13 +262,16 @@ class DistributedAPI:
             if self.local_client_arg is not None:
                 lc = local_client.LocalClient()
                 self.f_kwargs[self.local_client_arg] = lc
-
             try:
                 if self.is_async:
-                    task = run_local()
+                    task = self.run_local(self.f, self.f_kwargs, self.logger, self.rbac_permissions, self.broadcasting,
+                                          self.nodes, self.current_user)
+
                 else:
-                    loop = asyncio.get_running_loop()
-                    task = loop.run_in_executor(threadpool, run_local)
+                    loop = asyncio.get_event_loop()
+                    task = loop.run_in_executor(process_pool, partial(self.run_local, self.f, self.f_kwargs,
+                                                                      self.logger, self.rbac_permissions,
+                                                                      self.broadcasting, self.nodes, self.current_user))
                 try:
                     data = await asyncio.wait_for(task, timeout=timeout)
                 except asyncio.TimeoutError:
@@ -429,6 +438,7 @@ class DistributedAPI:
                     kcopy = deepcopy(self.to_dict())
                     if agent_list is not None and set(self.f_kwargs) & {'agent_id', 'agent_list'}:
                         kcopy['f_kwargs']['agent_id' if 'agent_id' in kcopy['f_kwargs'] else 'agent_list'] = agent_list
+
                     result = json.loads(await client.execute(b'dapi_fwd',
                                                              "{} {}".format(node_name,
                                                                             json.dumps(kcopy,
@@ -459,12 +469,12 @@ class DistributedAPI:
 
         async def clean_valid_nodes(nodes_to_clean: List[Tuple]) -> List[Tuple]:
             """Clean nodes response to forward only to real nodes in a single petition for each one.
-    
+
             Parameters
             ----------
             nodes_to_clean : list
                 List of nodes to clean.
-    
+
             Returns
             -------
             list

--- a/framework/wazuh/core/cluster/dapi/dapi.py
+++ b/framework/wazuh/core/cluster/dapi/dapi.py
@@ -32,6 +32,8 @@ from wazuh.core.exception import WazuhException, WazuhClusterError, WazuhError
 from wazuh.core.wazuh_socket import wazuh_sendsync
 
 process_pool = ProcessPoolExecutor(max_workers=1)
+authentication_pool = ProcessPoolExecutor(max_workers=1)
+authentication_funcs = {'check_token', 'check_user_master', 'get_permissions', 'get_security_conf'}
 
 
 class DistributedAPI:
@@ -225,6 +227,7 @@ class DistributedAPI:
     @staticmethod
     def run_local(f, f_kwargs, logger, rbac_permissions, broadcasting, nodes, current_user):
         """Run framework SDK function locally in another process."""
+
         def debug_log(logger, message):
             if logger.name == 'wazuh-api':
                 logger.debug2(message)
@@ -269,9 +272,15 @@ class DistributedAPI:
 
                 else:
                     loop = asyncio.get_event_loop()
-                    task = loop.run_in_executor(process_pool, partial(self.run_local, self.f, self.f_kwargs,
-                                                                      self.logger, self.rbac_permissions,
-                                                                      self.broadcasting, self.nodes, self.current_user))
+                    if self.f.__name__ not in authentication_funcs:
+                        pool = process_pool
+                    else:
+                        pool = authentication_pool
+
+                    task = loop.run_in_executor(pool, partial(self.run_local, self.f, self.f_kwargs,
+                                                              self.logger, self.rbac_permissions,
+                                                              self.broadcasting, self.nodes,
+                                                              self.current_user))
                 try:
                     data = await asyncio.wait_for(task, timeout=timeout)
                 except asyncio.TimeoutError:

--- a/framework/wazuh/core/cluster/dapi/dapi.py
+++ b/framework/wazuh/core/cluster/dapi/dapi.py
@@ -10,7 +10,7 @@ import operator
 import os
 import time
 from collections import defaultdict
-from concurrent.futures import ProcessPoolExecutor
+from concurrent.futures import process, ProcessPoolExecutor
 from copy import copy, deepcopy
 from functools import reduce, partial
 from operator import or_
@@ -28,6 +28,7 @@ from wazuh.cluster import get_node_wrapper, get_nodes_info
 from wazuh.core import common, exception
 from wazuh.core.cluster import local_client, common as c_common
 from wazuh.core.cluster.cluster import check_cluster_status
+from wazuh.core.cluster.common import WazuhJSONEncoder, as_wazuh_object
 from wazuh.core.exception import WazuhException, WazuhClusterError, WazuhError
 from wazuh.core.wazuh_socket import wazuh_sendsync
 
@@ -239,10 +240,14 @@ class DistributedAPI:
         common.broadcast.set(broadcasting)
         common.cluster_nodes.set(nodes)
         common.current_user.set(current_user)
-        data = f(**f_kwargs)
+        try:
+            data = f(**f_kwargs)
+        except Exception as e:
+            data = e
+
         common.reset_context_cache()
         debug_log(logger, "Finished executing request locally")
-        return data
+        return json.dumps(data, cls=WazuhJSONEncoder)
 
     async def execute_local_request(self) -> str:
         """Execute an API request locally.
@@ -282,13 +287,20 @@ class DistributedAPI:
                                                               self.broadcasting, self.nodes,
                                                               self.current_user))
                 try:
-                    data = await asyncio.wait_for(task, timeout=timeout)
+                    data = json.loads(await asyncio.wait_for(task, timeout=timeout), object_hook=as_wazuh_object)
+                    if isinstance(data, Exception):
+                        raise data
+
                 except asyncio.TimeoutError:
                     raise exception.WazuhInternalError(3021)
                 except OperationalError:
                     raise exception.WazuhInternalError(2008)
+                except process.BrokenProcessPool:
+                    raise exception.WazuhInternalError(901)
             except json.decoder.JSONDecodeError:
                 raise exception.WazuhInternalError(3036)
+            except process.BrokenProcessPool:
+                raise exception.WazuhInternalError(900)
 
             self.debug_log(f"Time calculating request result: {time.time() - before:.3f}s")
             return data

--- a/framework/wazuh/core/cluster/master.py
+++ b/framework/wazuh/core/cluster/master.py
@@ -706,7 +706,7 @@ class MasterHandler(server.AbstractServerHandler, c_common.WazuhCommon):
                 task_id = await self.send_request(command=b'syn_m_c', data=b'')
                 if isinstance(task_id, Exception) or task_id.startswith(b'Error'):
                     exc_info = task_id if isinstance(task_id, Exception) else \
-                        exception.WazuhClusterError(code=3016, extra_message=str(task_id))
+                        exception.WazuhClusterError(3016, extra_message=str(task_id))
                     task_id = b'None'
                     raise exc_info
 
@@ -729,7 +729,7 @@ class MasterHandler(server.AbstractServerHandler, c_common.WazuhCommon):
             except Exception as e:
                 # Notify error to worker and delete its received file.
                 self.logger.error(f"Error sending files information: {e}")
-                exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+                exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                       cls=c_common.WazuhJSONEncoder).encode()
                 result = await self.send_request(command=b'syn_m_c_r', data=task_id + b' ' + exc_info)
             finally:

--- a/framework/wazuh/core/cluster/server.py
+++ b/framework/wazuh/core/cluster/server.py
@@ -306,7 +306,7 @@ class AbstractServer:
             select = default_fields
         else:
             if not set(select).issubset(default_fields):
-                raise exception.WazuhError(code=1724, extra_message=', '.join(set(select) - default_fields),
+                raise exception.WazuhError(1724, extra_message=', '.join(set(select) - default_fields),
                                            extra_remediation=', '.join(default_fields))
 
         if filter_type != 'all' and filter_type not in {'worker', 'master'}:

--- a/framework/wazuh/core/cluster/tests/test_common.py
+++ b/framework/wazuh/core/cluster/tests/test_common.py
@@ -19,10 +19,10 @@ with patch('wazuh.common.getgrnam'):
                 from wazuh.core.cluster.common import WazuhJSONEncoder, as_wazuh_object
 
 affected = AffectedItemsWazuhResult(dikt={'data': ['test']}, affected_items=['001', '002'])
-affected.add_failed_item(id_='099', error=WazuhError(code=1750, extra_message='wiiiiiii'))
-affected.add_failed_item(id_='111', error=WazuhError(code=1750, extra_message='weeeeee'))
-affected.add_failed_item(id_='555', error=WazuhError(code=1750, extra_message='wiiiiiii'))
-affected.add_failed_item(id_='333', error=WazuhError(code=1707, extra_message='wiiiiiii'))
+affected.add_failed_item(id_='099', error=WazuhError(1750, extra_message='wiiiiiii'))
+affected.add_failed_item(id_='111', error=WazuhError(1750, extra_message='weeeeee'))
+affected.add_failed_item(id_='555', error=WazuhError(1750, extra_message='wiiiiiii'))
+affected.add_failed_item(id_='333', error=WazuhError(1707, extra_message='wiiiiiii'))
 
 with patch('wazuh.common.wazuh_path', new=os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data')):
     objects_to_encode = [

--- a/framework/wazuh/core/cluster/worker.py
+++ b/framework/wazuh/core/cluster/worker.py
@@ -123,7 +123,7 @@ class SyncFiles(SyncTask):
             raise task_id
         elif task_id.startswith(b'Error'):
             self.logger.error(task_id.decode())
-            exc_info = json.dumps(exception.WazuhClusterError(code=3016, extra_message=str(task_id)),
+            exc_info = json.dumps(exception.WazuhClusterError(3016, extra_message=str(task_id)),
                                   cls=c_common.WazuhJSONEncoder).encode()
             await self.worker.send_request(command=self.cmd + b'_r', data=b'None ' + exc_info)
             return
@@ -156,7 +156,7 @@ class SyncFiles(SyncTask):
         except Exception as e:
             # Notify error to master and delete its received file.
             self.logger.error(f"Error sending zip file: {e}")
-            exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+            exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                   cls=c_common.WazuhJSONEncoder).encode()
             await self.worker.send_request(command=self.cmd+b'_r', data=task_id + b' ' + exc_info)
         finally:
@@ -620,7 +620,7 @@ class WorkerHandler(client.AbstractClient, c_common.WazuhCommon):
                                         json.dumps(e, cls=c_common.WazuhJSONEncoder).encode())
             except Exception as e:
                 logger.error(f"Error synchronizing integrity: {e}")
-                exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+                exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                       cls=c_common.WazuhJSONEncoder)
                 await self.send_request(command=b'syn_i_w_m_r', data=b'None ' + exc_info.encode())
 
@@ -693,7 +693,7 @@ class WorkerHandler(client.AbstractClient, c_common.WazuhCommon):
                                     data=b'None ' + json.dumps(e, cls=c_common.WazuhJSONEncoder).encode())
         except Exception as e:
             logger.error(f"Error synchronizing extra valid files: {e}")
-            exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+            exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                   cls=c_common.WazuhJSONEncoder)
             await self.send_request(command=b'syn_i_w_m_r', data=b'None ' + exc_info.encode())
 
@@ -715,7 +715,7 @@ class WorkerHandler(client.AbstractClient, c_common.WazuhCommon):
 
         if isinstance(self.sync_tasks[name].filename, Exception):
             exc_info = json.dumps(exception.WazuhClusterError(
-                code=1000, extra_message=str(self.sync_tasks[name].filename)), cls=c_common.WazuhJSONEncoder)
+                1000, extra_message=str(self.sync_tasks[name].filename)), cls=c_common.WazuhJSONEncoder)
             await self.send_request(command=b'syn_i_w_m_r', data=b'None ' + exc_info.encode())
             raise self.sync_tasks[name].filename
 
@@ -759,7 +759,7 @@ class WorkerHandler(client.AbstractClient, c_common.WazuhCommon):
                                     data=b'None ' + json.dumps(e, cls=c_common.WazuhJSONEncoder).encode())
         except Exception as e:
             logger.error(f"Error synchronizing extra valid files: {e}")
-            exc_info = json.dumps(exception.WazuhClusterError(code=1000, extra_message=str(e)),
+            exc_info = json.dumps(exception.WazuhClusterError(1000, extra_message=str(e)),
                                   cls=c_common.WazuhJSONEncoder)
             await self.send_request(command=b'syn_i_w_m_r', data=b'None ' + exc_info.encode())
         finally:

--- a/framework/wazuh/core/common.py
+++ b/framework/wazuh/core/common.py
@@ -10,6 +10,7 @@ from copy import deepcopy
 from functools import lru_cache
 from functools import wraps
 from grp import getgrnam
+from multiprocessing import Event
 from pwd import getpwnam
 from typing import Dict, Any
 
@@ -222,6 +223,9 @@ cluster_integrity_mtime: ContextVar[Dict] = ContextVar('cluster_integrity_mtime'
 origin_module: ContextVar[str] = ContextVar('origin_module', default='framework')
 
 _context_cache = dict()
+
+# Clear cache event
+cache_event = Event()
 
 
 def context_cached(key: str = '') -> Any:

--- a/framework/wazuh/core/exception.py
+++ b/framework/wazuh/core/exception.py
@@ -17,6 +17,9 @@ class WazuhException(Exception):
 
     ERRORS = {
         # < 999: API
+        900: 'One of the API child processes terminated abruptly. The API process pool is not usable anymore. '
+             'Please restart the Wazuh API',
+        901: 'API executor subprocess broke. A service restart may be needed',
 
         # Wazuh: 0999 - 1099
         999: 'Incompatible version of Python',

--- a/framework/wazuh/core/exception.py
+++ b/framework/wazuh/core/exception.py
@@ -232,9 +232,9 @@ class WazuhException(Exception):
                },
         1707: {'message': 'Cannot send request, agent is not active',
                'remediation': 'Please, check non-active agents connection and try again. Visit '
-               f'https://documentation.wazuh.com/{WAZUH_VERSION}/user-manual/registering/index.html and '
-               f'https://documentation.wazuh.com/{WAZUH_VERSION}/user-manual/agents/agent-connection.html'
-               ' to obtain more information on registering and connecting agents'
+                              f'https://documentation.wazuh.com/{WAZUH_VERSION}/user-manual/registering/index.html and '
+                              f'https://documentation.wazuh.com/{WAZUH_VERSION}/user-manual/agents/agent-connection.html'
+                              ' to obtain more information on registering and connecting agents'
                },
         1708: {'message': 'There is an agent with the same ID',
                'remediation': 'Please choose another ID'
@@ -560,6 +560,11 @@ class WazuhException(Exception):
 
         # > 9000: Authd
     }
+
+    # Reserve agent upgrade custom errors
+    ERRORS.update({key: {'message': 'Upgrade module\'s reserved exception IDs (1810-1899). '
+                                    'The error message will be the output of upgrade module'}
+                   for key in range(1811, 1900)})
 
     def __init__(self, code, extra_message=None, extra_remediation=None, cmd_error=False, dapi_errors=None, title=None,
                  type=None):

--- a/framework/wazuh/logtest.py
+++ b/framework/wazuh/logtest.py
@@ -39,7 +39,7 @@ def run_logtest(token=None, event=None, log_format=None, location=None):
 
     response = send_logtest_msg(command='log_processing', parameters=local_vars)
     if response['error'] != 0:
-        raise WazuhError(code=7000, extra_message=response.get('message', 'Could not parse error message'))
+        raise WazuhError(7000, extra_message=response.get('message', 'Could not parse error message'))
 
     return response
 
@@ -63,6 +63,6 @@ def end_logtest_session(token: str = None):
 
     response = send_logtest_msg(command='remove_session', parameters={'token': token})
     if response['error'] != 0:
-        raise WazuhError(code=7000, extra_message=response.get('message', 'Could not parse error message'))
+        raise WazuhError(7000, extra_message=response.get('message', 'Could not parse error message'))
 
     return response

--- a/framework/wazuh/rbac/preprocessor.py
+++ b/framework/wazuh/rbac/preprocessor.py
@@ -123,7 +123,7 @@ def get_permissions(user_id=None, auth_context=None):
     """
     with AuthenticationManager() as auth:
         if not auth.user_allow_run_as(user_id) and auth_context:
-            raise WazuhPermissionError(code=6004)
+            raise WazuhPermissionError(6004)
         elif auth.user_allow_run_as(user_id):
             roles = get_roles(auth_context=auth_context, user_id=user_id)
         else:

--- a/framework/wazuh/rbac/utils.py
+++ b/framework/wazuh/rbac/utils.py
@@ -5,6 +5,7 @@
 from functools import wraps
 
 from cachetools import TTLCache, cached
+from wazuh.core.common import cache_event
 
 from api.configuration import security_conf
 
@@ -14,7 +15,7 @@ tokens_cache = TTLCache(maxsize=4500, ttl=security_conf['auth_token_exp_timeout'
 
 def clear_cache():
     """This function clear the authorization tokens cache."""
-    tokens_cache.clear()
+    cache_event.set()
 
 
 def token_cache(cache):
@@ -33,6 +34,10 @@ def token_cache(cache):
         @wraps(func)
         def wrapper(*args, **kwargs):
             origin_node_type = kwargs.pop('origin_node_type')
+
+            if cache_event.is_set():
+                cache.clear()
+                cache_event.clear()
 
             @cached(cache=cache)
             def f(*_args, **_kwargs):

--- a/framework/wazuh/security.py
+++ b/framework/wazuh/security.py
@@ -6,11 +6,11 @@ import re
 from copy import deepcopy
 from functools import lru_cache
 
-import api.configuration as configuration
+from api.authentication import get_security_conf
 from wazuh.core import common
 from wazuh.core.exception import WazuhError, WazuhResourceNotFound
 from wazuh.core.results import AffectedItemsWazuhResult, WazuhResult
-from wazuh.core.security import invalid_users_tokens, invalid_roles_tokens, invalid_run_as_tokens, revoke_tokens,\
+from wazuh.core.security import invalid_users_tokens, invalid_roles_tokens, invalid_run_as_tokens, revoke_tokens, \
     load_spec, sanitize_rbac_policy, update_security_conf, REQUIRED_FIELDS, SORT_FIELDS, SORT_FIELDS_GET_USERS
 from wazuh.core.utils import process_array
 from wazuh.rbac.decorators import expose_resources
@@ -1103,7 +1103,7 @@ def get_rbac_actions(endpoint: str = None):
 @expose_resources(actions=['security:read_config'], resources=['*:*:*'])
 def get_security_config():
     """Returns current security configuration."""
-    return WazuhResult({'data': configuration.security_conf})
+    return WazuhResult({'data': get_security_conf()})
 
 
 @expose_resources(actions=['security:update_config'], resources=['*:*:*'])

